src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs

```diff
@@ -1298,32 +1311,57 @@ private void OnHeaderCore(int? index, bool indexedValue, ReadOnlySpan<byte> name
             throw new Http2ConnectionErrorException(CoreStrings.BadRequest_HeadersExceedMaxTotalSize, Http2ErrorCode.PROTOCOL_ERROR);
         }
 
-        if (index != null)
-        {
-            ValidateStaticHeader(index.Value, value);
-        }
-        else
-        {
-            ValidateHeader(name, value);
-        }
-
         try
         {
             if (_requestHeaderParsingState == RequestHeaderParsingState.Trailers)
             {
+                UpdateHeaderParsingState(value, GetPseudoHeaderField(name));
+                ValidateHeaderContent(name, value);
```

Comment:
Does this assume trailers aren't dynamically compressed? or just that they're uncommon enough not to matter?

Reply:
The logic around checking whether a header is compressed or not is all optimization. Given how infrequent they are (is anyone using them) I didn't think it was worth doing for trailing headers on requests.

I'll add a comment saying there is the opportunity for improvement.